torch/_functorch/aot_autograd.py

```diff
@@ -1648,6 +1648,83 @@ def same_dtype_views(view1, view2):
     return True
 
 
+
+# Assumption: x and y are known to share a storage, and we are trying to determine
+# if their memory is actually completely disjoint, based on sizes/strides/storage_offset
+def tensors_definitely_do_not_overlap(x, y):
+    if x is y:
+        return False
+    if x.numel() == 0 or y.numel() == 0:
+        return True
+
+    # Make x always on the left
+    if x.storage_offset() > y.storage_offset():
+        x, y = y, x
+    # Short-circuit in the "obvious" overlapping case: both tensors are contiguous
+    if x.is_contiguous() and y.is_contiguous():
+        if x.storage_offset() + x.numel() > y.storage_offset():
+            # definitely overlap
+            return False
+        else:
+            # definitely no overlap
+            return True
+
+    if x.dim() == 2 and y.dim() == 2 and x.stride(1) == 1 and y.stride(1) == 1:
+        # This cases is needed for the shampoo optimizer.
+        # All tensors are 2d (non-contiguous), have the same outer stride, and have an inner stride of 1
+        # (so rows are contiguous)
+        if x.stride(0) == y.stride(0):
+            offset_delta = y.storage_offset() - x.storage_offset()
+            if offset_delta < x.size(1):
+                # definitely overlaps (row 0 of y overlaps with row 0 of x)
+                # Example:
+                #   x: size=(4, 4), stride=(8, 1), offset=0
+                #   y: size=(4, 4), stride=(8, 1), offset=3
```

Comment:
nit: I would mention that you get these as `x = base[:4]` and `y = base[3:]`. That makes it easier to visualize for me.

Reply:
Hmm... is there actually a nice set of torch view ops that represents the strides and storage_offsets in my example? I don't think `base[:4]` and `base[3:]` will reflect the fact that x and y are non-contiguous (because they have a large outer stride)

Reply:
yep you're right, that's exactly it. will update